# pandas: load a CSV, then `df.describe()`

Two tasks from **Luis Junco**, Slack `#2--class-activities`, 17 Sept:

1. **11:30 - Practice: loading a csv file in pandas** (10 min)
   `avocado_kaggle.csv` from the class repo `demos_ai_engineering_sept2026`
2. **11:52 - Research: `df.describe()`** (8 min)
   notes in `C_pandas/1_pandas_intro.ipynb` of the same repo

The file is already downloaded to `01_python/data/avocado_kaggle.csv`.

---
## 1. Load the CSV

In [5]:
import pandas as pd

df = pd.read_csv("data/avocado_kaggle.csv", index_col=0)
# index_col=0 uses the file's first column (an old row counter) as the index,
# otherwise it shows up as a column called "Unnamed: 0"

print("rows, columns:", df.shape)
df.head()

rows, columns: (18249, 13)


,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015,Albany
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015,Albany
2,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015,Albany
3,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015,Albany
4,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015,Albany


### What is in it?

In [6]:
print(list(df.columns))
print()
print(df.dtypes)

['Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year', 'region']

Date                str
AveragePrice    float64
Total Volume    float64
4046            float64
4225            float64
4770            float64
Total Bags      float64
Small Bags      float64
Large Bags      float64
XLarge Bags     float64
type                str
year              int64
region              str
dtype: object


In [7]:
df.info()      # column names, non-null counts, types, memory use

<class 'pandas.DataFrame'>
Index: 18249 entries, 0 to 11
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          18249 non-null  str    
 1   AveragePrice  18249 non-null  float64
 2   Total Volume  18249 non-null  float64
 3   4046          18249 non-null  float64
 4   4225          18249 non-null  float64
 5   4770          18249 non-null  float64
 6   Total Bags    18249 non-null  float64
 7   Small Bags    18249 non-null  float64
 8   Large Bags    18249 non-null  float64
 9   XLarge Bags   18249 non-null  float64
 10  type          18249 non-null  str    
 11  year          18249 non-null  int64  
 12  region        18249 non-null  str    
dtypes: float64(9), int64(1), str(3)
memory usage: 1.9 MB


---
## 2. `df.describe()` - the statistical summary

One call, and every **numeric** column gets 8 numbers.

In [8]:
df.describe()

,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,year
count,18249.000000,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,18249.000000,18249.000000
mean,1.405978,8.506440e+05,2.930084e+05,2.951546e+05,2.283974e+04,2.396392e+05,1.821947e+05,5.433809e+04,3106.426507,2016.147899
std,0.402677,3.453545e+06,1.264989e+06,1.204120e+06,1.074641e+05,9.862424e+05,7.461785e+05,2.439660e+05,17692.894652,0.939938
min,0.440000,8.456000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,2015.000000
25%,1.100000,1.083858e+04,8.540700e+02,3.008780e+03,0.000000e+00,5.088640e+03,2.849420e+03,1.274700e+02,0.000000,2015.000000
50%,1.370000,1.073768e+05,8.645300e+03,2.906102e+04,1.849900e+02,3.974383e+04,2.636282e+04,2.647710e+03,0.000000,2016.000000
75%,1.660000,4.329623e+05,1.110202e+05,1.502069e+05,6.243420e+03,1.107834e+05,8.333767e+04,2.202925e+04,132.500000,2017.000000
max,3.250000,6.250565e+07,2.274362e+07,2.047057e+07,2.546439e+06,1.937313e+07,1.338459e+07,5.719097e+06,551693.650000,2018.000000


### The same thing for a single column (a Series)

In [9]:
df["AveragePrice"].describe()

count    18249.000000
mean         1.405978
std          0.402677
min          0.440000
25%          1.100000
50%          1.370000
75%          1.660000
max          3.250000
Name: AveragePrice, dtype: float64

### What each row means

| row | meaning |
|---|---|
| `count` | how many **non-empty** values the column has |
| `mean` | the average |
| `std` | standard deviation - how spread out the values are around the mean |
| `min` | the smallest value |
| `25%` | **Q1**: 25% of the values are below this |
| `50%` | the **median**: half the values are below, half above |
| `75%` | **Q3**: 75% of the values are below this |
| `max` | the largest value |

Read as a sentence for `AveragePrice`: *18249 prices, averaging 1.41 dollars, the cheapest
0.44 and the dearest 3.25; half of them are under 1.37, and the middle 50% sit between
1.10 and 1.66.*

`count` is the quiet one - if it is smaller than the number of rows, that column has
**missing values**.

---
## 3. Check describe() by hand

The same numbers, computed one by one. `std` uses **n - 1** (the sample version), exactly
like the `st_dev` we wrote in the functions lab.

In [10]:
price = df["AveragePrice"]

print("count :", price.count())
print("mean  :", round(price.mean(), 6))
print("std   :", round(price.std(), 6))        # ddof=1 by default -> divides by n-1
print("min   :", price.min())
print("25%   :", price.quantile(0.25))
print("50%   :", price.median(), "= quantile(0.5):", price.quantile(0.50))
print("75%   :", price.quantile(0.75))
print("max   :", price.max())

# numpy's std divides by n instead, so it differs slightly:
import numpy as np
print()
print("pandas std (n-1):", round(price.std(), 6))
print("numpy  std (n)  :", round(np.std(price.values), 6))

count : 18249
mean  : 1.405978
std   : 0.402677
min   : 0.44
25%   : 1.1
50%   : 1.37 = quantile(0.5): 1.37
75%   : 1.66
max   : 3.25

pandas std (n-1): 0.402677
numpy  std (n)  : 0.402666


---
## 4. describe() on the text columns

By default `describe()` ignores text. Ask for it and the rows change completely.

In [11]:
# include="object" also works, but pandas 3 warns about it:
# text columns are now the "str" dtype
df.describe(include="str")

,Date,type,region
count,18249,18249,18249
unique,169,2,54
top,2015-12-27,conventional,Albany
freq,108,9126,338


| row | meaning |
|---|---|
| `count` | how many values |
| `unique` | how many **different** values |
| `top` | the most frequent value |
| `freq` | how often the top value appears |

`include="all"` puts numbers and text in one table (with `NaN` where a statistic
does not apply).

In [12]:
df.describe(include="all")

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
count,18249,18249.000000,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,1.824900e+04,18249.000000,18249,18249.000000,18249
unique,169,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,NaN,54
top,2015-12-27,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,conventional,NaN,Albany
freq,108,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9126,NaN,338
mean,NaN,1.405978,8.506440e+05,2.930084e+05,2.951546e+05,2.283974e+04,2.396392e+05,1.821947e+05,5.433809e+04,3106.426507,NaN,2016.147899,NaN
std,NaN,0.402677,3.453545e+06,1.264989e+06,1.204120e+06,1.074641e+05,9.862424e+05,7.461785e+05,2.439660e+05,17692.894652,NaN,0.939938,NaN
min,NaN,0.440000,8.456000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,NaN,2015.000000,NaN
25%,NaN,1.100000,1.083858e+04,8.540700e+02,3.008780e+03,0.000000e+00,5.088640e+03,2.849420e+03,1.274700e+02,0.000000,NaN,2015.000000,NaN
50%,NaN,1.370000,1.073768e+05,8.645300e+03,2.906102e+04,1.849900e+02,3.974383e+04,2.636282e+04,2.647710e+03,0.000000,NaN,2016.000000,NaN
75%,NaN,1.660000,4.329623e+05,1.110202e+05,1.502069e+05,6.243420e+03,1.107834e+05,8.333767e+04,2.202925e+04,132.500000,NaN,2017.000000,NaN


---
## 5. Two useful variations

In [13]:
# your own percentiles instead of 25/50/75
df["AveragePrice"].describe(percentiles=[0.1, 0.5, 0.9])

count    18249.000000
mean         1.405978
std          0.402677
min          0.440000
10%          0.930000
50%          1.370000
90%          1.930000
max          3.250000
Name: AveragePrice, dtype: float64

In [14]:
# describe() AFTER filtering - the summary follows the rows you kept
organic = df[df["type"] == "organic"]
conventional = df[df["type"] == "conventional"]

print("organic     :", len(organic), "rows, mean price", round(organic["AveragePrice"].mean(), 3))
print("conventional:", len(conventional), "rows, mean price", round(conventional["AveragePrice"].mean(), 3))
print()
organic["AveragePrice"].describe()

organic     : 9123 rows, mean price 1.654
conventional: 9126 rows, mean price 1.158



count    9123.000000
mean        1.653999
std         0.363502
min         0.440000
25%         1.420000
50%         1.630000
75%         1.870000
max         3.250000
Name: AveragePrice, dtype: float64

In [15]:
# the same comparison in one line
df.groupby("type")["AveragePrice"].describe()

,count,mean,std,min,25%,50%,75%,max
type,,,,,,,,
conventional,9126.0,1.158040,0.263041,0.46,0.98,1.13,1.32,2.22
organic,9123.0,1.653999,0.363502,0.44,1.42,1.63,1.87,3.25


---
## Remember

- `df.describe()` -> numeric columns only; `include="object"` -> text; `include="all"` -> both
- `50%` **is** the median, and `25%`/`75%` mark the middle half of the data
- pandas `std` divides by **n - 1**, numpy's `np.std` divides by **n**
- `count` lower than `df.shape[0]` is your first hint of missing values